typecheck-requirements.txt

```diff
@@ -0,0 +1,5 @@
+# Contains libraries or stub requirements which are used in the typechecked code. Pyright needs
+# these available to check the code.
+
+# Used by instrumentation-genai/opentelemetry-instrumentation-vertexai
```

Comment:
I guess we'll just keep adding to this file more packages opt in to typechecking. I couldn't come up with a more automated way to achieve this.

Maybe we can use the `uv.lock` file from https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3124?

Reply:
What do people use the `dev-requirements.txt` for? 🤔 Can we add this there?

Yeah, we should use `[dependency-groups]` on the pyproject to simplify the installation of dependencies. When that PR is merged, we can simplify the requirements files.

Reply:
> What do people use the `dev-requirements.txt` for? 🤔 Can we add this there?

It's used in the tox file a few places to install tooling e.g. in `tox -e ruff`. This bloats things a bit since it pulls in all the transitive deps, especially once we have more libraries being typechecked:

```console
$ du -sh .tox/ruff/
12M	.tox/ruff/
$ du -sh .tox/typecheck/
164M	.tox/typecheck/
```

Reply:
That's heavy... 👀

Reply:
I don't like adding new `requirements.txt`, but all good. 👍

Reply:
Actually, can you do this instead on the tox.ini typecheck:
```
  {toxinidir}/instrumentation-genai/opentelemetry-instrumentation-vertexai[instruments]
```
I'll organize the dependencies on the main pyproject when the uv PR is merged.

Reply:
This would also install the package itself. Does Pyright behave correctly when the code your are type checking is also installed into the virtualenv?

Reply:
I updated to do this but still curious your thoughts. I know that Mypy can behave differently if the code being checked is installed in virtualenv vs. specified as arguments vs. added to PYTHONPATH

Reply:
That should be seen as a package anyway. If other package internally references (imports) that package, it should consider it as it is locally.

We want local type checked folders to see other folders in the repository as packages.

Reply:
OK I'm not really sure how pyright is supposed to work with monorepos. I saw some evidence in the docs that it understands `src/` layouts.

But this seems to be working for now so we can revisit if needed 👍